app/models/posts.js

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+export class Posts {
+    get propsAsJson() {
+        try {
+            return JSON.parse(this.props);
+        } catch {
+            return null;
+        }
+    }
+
+    set propsFromJson(props) {
+        this.props = JSON.stringify(props);
+    }
+
+    static schema = {
+        name: 'Posts',
+        primaryKey: 'id',
+        properties: {
+            id: 'string',
+            createAt: {type: 'int', indexed: true},
+            updateAt: {type: 'int', indexed: true},
+            deleteAt: {type: 'int', default: 0},
+            editAt: {type: 'int', default: 0},
+            user: 'Users',
+            rootId: 'string?',
+            originalId: 'string',
+            pendingPostId: 'string?',
+            message: {type: 'string', default: ''},
+            type: 'string?',
+            props: 'string',
+            files: 'Files[]',
+            hasReactions: {type: 'bool', default: false},
+            isPinned: {type: 'bool', default: false},
+            metadata: 'Metadata[]',
+            reactions: 'Reactions[]',
+        },
+    }
+}
+
+export class Metadata {
+    get dataAsJson() {
+        try {
+            return JSON.parse(this.data);
+        } catch {
+            return null;
+        }
+    }
+
+    static schema = {
+        name: 'Metadata',
+        primaryKey: 'hash',
+        properties: {
+            hash: 'string',
```

Comment:
The client doesn't need to store the hash for link metadata. That's only used as a unique ID on the server since it's derived from the URL and timestamp